File: Interviewbits/Matrix Search.cpp
int binarySearch(vector<int> Arr,int target) {

    //set stating and ending index
    int start = 0, end = Arr.size()-1;

    while(start <= end) {
        // take mid of the list
        int mid = (start + end) / 2;

        // we found a match
        if(Arr[mid] == target) {
            return 1; 
        }
        // go on right side
        else if(Arr[mid] < target) {
            start = mid + 1;
        }
        // go on left side
        else {
            end = mid - 1;
        }
    }
    // element is not present in list
    return 0;
}

int Solution::searchMatrix(vector<vector<int> > &A, int B) {
    int r=A.size();
    int c=A[0].size();
   for(int i=0;i<r;i++){
        if(A[i][0]<=B && A[i][c-1]>=B){
            return binarySearch(A[i],B);
        }
    }
    return 0;
}

File: Interviewbits/Level3-Binary-Search/Matrix Search.py
def binarySearch(arr, l, r, x): 
  
    while l <= r: 
  
        mid = l + (r - l)/2; 
          
        # Check if x is present at mid 
        if arr[mid] == x: 
            return mid 
  
        # If x is greater, ignore left half 
        elif arr[mid] < x: 
            l = mid + 1
  
        # If x is smaller, ignore right half 
        else: 
            r = mid - 1
      
    # If we reach here, then the element 
    # was not present 
    return -1
def searchMatrix(A, B):
    # A->Matrix
    # B->Integer

def main():
    # *****************Input********************
    [r,c]=list(map(int,input().split()))
    A=[]
    for i in range(r):
        temp=list(map(int,input().split()))
        A.append(temp)
    search=int(input())
    # *****************Algo*********************
    answer=searchMatrix(A,search)

    # *****************Output*******************
    print(answer)
    
main()

